play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/ArscTable.java

```diff
@@ -0,0 +1,129 @@
+     * @param aabPath The path to the root of the AAB directory.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAabDirectory(Path aabPath) throws IOException {
+        Path arscPath = Paths.get(aabPath.toString(), RESOURCES_FILE_NAME);
+        File arscFile = new File(arscPath.toAbsolutePath().toString());
+        if (!arscFile.exists()) {
+            throw new FileNotFoundException("Resources file not found");
+        }
+        InputStream is = new FileInputStream(arscFile);
+        return createTable(is, null, aabPath);
+    }
+
+    /**
+     * Creates the table from a AAB or APK file.
+     * @param zipFile The zip file object representing the AAB/APK.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAndroidPackage(ZipFile zipFile) throws IOException {
+        ZipEntry arscEntry = new ZipEntry(RESOURCES_FILE_NAME);
+        InputStream is = zipFile.getInputStream(arscEntry);
+        return createTable(is, zipFile, null);
+    }
+
+    /**
+     * Creates the table from an InputStream.
+     *
+     * @param is The InputStream
+     * @param zipFile The source zip file. This must be populated when creating the table from a AAB
+     *                or APK archive. Where the filesystem approach is taken, this should be null.
+     * @param aabRoot The source AAB directory path. This should be null when creating the table
+     *                from an AAB/APK archive.
+     * @return The constructed table.
+     * @throws IOException when errors loading resources occur.
+     */
+    private static ArscTable createTable(InputStream is, ZipFile zipFile, Path aabRoot) throws IOException {
+        BinaryResourceFile resources = BinaryResourceFile.fromInputStream(is);
+        ResourceTableChunk table = (ResourceTableChunk) resources.getChunks().get(0);
+        StringPoolChunk stringPool = table.getStringPool();
+
+        List<TypeChunk> chunks = table.getPackages()
+                .stream()
+                .flatMap(p -> p.getTypeChunks().stream())
+                .toList();
+
+        List<ArscResource> resourcesList = chunks.stream().flatMap(c -> c.getEntries().values().stream())
+                .filter(t -> RESOURCE_TYPES.contains(t.typeName()))
+                .filter(t -> t.value().type() == BinaryResourceValue.Type.STRING)
+                .map(entry -> {
+                    Path path = Path.of(stringPool.getString(entry.value().data()));
+                    byte[] data = null;
+                    try {
+                        // Loading the resource data requires a different approach depending on
+                        // whether the data is in a zip file or directly on the file system.
+                        if (zipFile != null) {
```

Comment:
You could uplift this logic out of createTable, and provide a lambda specific to createFrom*. Avoids a branch here, keeps it simpler.

#tioli

Reply:
Ok, I think I've done that.... my Java is rusty.